### PR TITLE
[WIP] Refactor casper code to avoid the usage of Option fields in proto classes

### PR DESCRIPTION
**Pull request by itegulov**
## Overview
Working with proto classes in their current can be quite tedious as you have to unwrap optional fields (e.g. body, header) which aren't actually optional after validating. The proposed solution is to generate safe proto classes based on their implementation provided by scalapb. This is done using a single macro `@GenerateSafe` which generates a safe version of the annotated class. All optional fields become non-optional in the safe version. Also, the underlying type is converted to its safe version. For example, the parameter `body: Option[Body]` in `BlockMessage` will be converted to `body: BodySafe`. This way we can ensure that the entire data structure is safe to work with, not only its immediate parameters. Along with the safe parameters, a handful of helper methods are generated for the safe version of a class: `toByteString`, `toByteArray`, `with<paramname>`.

However, there are a few problems with the current implementation:

1. ~~As it is impossible to generate non-eponymous top-level classes (e.g. `BodySafe` for `Body`) from an annotated class, I have decided to put a safe version of the class in the companion object. For example, the current implementation  generates `Body.BodySafe` for `Body`, but I might change it to just `Body.Safe` as the current version is redundant and really hard to work with right now.~~ **UPD**: Now generated classes are named like `Body.Safe` instead of `Body.BodySafe`.
2. I know that the experience working with the codebase relying on macros can be less than pleasant. So I have implemented [an extension for IntelliJ IDEA Scala plugin](https://github.com/itegulov/rchain-idea). You shouldn't have any problems with macros if you are using ENSIME.

Any feedback is welcome as it is a big change affecting most of the files in casper, block-storage and comm.
   
### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
https://rchain.atlassian.net/browse/RHOL-572

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You assigned one person to review this PR

### CODE REVIEW

**Comment by goral09** [comment]
Thanks @itegulov , this looks really promising. I haven't gone through the whole code, I just read the description, I have a question though -- we expect that in the future there will be many 3rd party clients integrating with the node, using the protobuf as API. How can we ensure that those not-yet-existing clients won't send us data with empty fields?

**Comment by itegulov** [comment]
@goral09, I don't think we can ensure that those clients won't send us data with empty fields. The original `BlockMessage` is still here, but after validating its fields (using `Validate.formatOfFields`) we can safely convert it to `BlockMessage.BlockMessageSafe`. And all further work is done only using the safe instance of the message.

**Comment by MParlikar** [comment]
Let's please focus on getting this reviewed. This PR now has conflicts that need to be resolved.

File: casper/src/main/scala/coop/rchain/casper/Casper.scala [hunk 1]
```diff
@@ -161,8 +162,31 @@ sealed abstract class MultiParentCasperInstances {
...
+                                    case Right(blockSafe) =>
+                                      for {
+                                        isOldBlock <- contains(blockSafe)
+                                        result <- if (isOldBlock) {
+                                                   for {
+                                                     _ <- Log[F].info(
+                                                           s"Received block ${PrettyPrinter
+                                                             .buildString(b.blockHash)} again.")
+                                                     result <- Monad[F].pure[BlockStatus](
+                                                                InvalidUnslashableBlock)
```
Comment (birchmd) [inline casper/src/main/scala/coop/rchain/casper/Casper.scala]: I don't really like that a block we already have is considered invalid. Maybe make a new status to cover this case.

File: casper/src/main/scala/coop/rchain/casper/Validate.scala [hunk 2]
```diff
@@ -106,49 +109,55 @@ object Validate {
...
       for {
         _ <- Log[F].warn(ignore(b, s"one of block comm reduction events is empty."))
-      } yield false
+      } yield Left(InvalidUnslashableBlock)
     } else {
-      true.pure[F]
+      BlockMessage.BlockMessageSafe
+        .create(b)
+        .getOrElse(throw new IllegalStateException("Block is not safe after format validation"))
+        .asRight[InvalidBlock]
```
[the remark this answers is not in the document]
Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala]: No, this lifts the result to a `Right` and the type parameter gives the type for the `Left` side of the `Either`. See https://typelevel.org/cats/api/cats/syntax/EitherIdOps.html#asRight[B]:Either[B,A]

[the remark this answers is not in the document]
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala]: Ah yes the unfortunately necessary asLeft/asRight construct.

File: casper/src/main/scala/coop/rchain/casper/Casper.scala (hunk 1 above)
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Casper.scala]: Can we extract this into a new method? It has become hard to read for me.

**Comment by goral09** [comment]
@itegulov how do I use the intellij plugin?

**Comment by itegulov** [comment]
@KentShikama It might make sense to have a general option to generate safe proto classes in scalapb, but they could just generate the code for safe versions (along with the currently generated code) instead of generating them by using macros. I will reach out to the maintainers of scalapb and discuss with them whether they want this feature in scalapb.

**Comment by guardbotmk3** [comment]
I don't understand why I'm assigned to this PR

**Pull request closed without merging**